csi/controller.go

```diff
@@ -455,9 +471,122 @@ func csiRequestsSharedVolume(req *csi.CreateVolumeRequest) bool {
 	return false
 }
 
-/*
-For next patches what still needs to be worked on in the Conroller server:
+// CreateSnapshot is a CSI implementation to create a snapshot from the volume
+func (o *OsdCsiServer) CreateSnapshot(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
 
-	GetCapacity(context.Context, *GetCapacityRequest) (*GetCapacityResponse, error)
-	ControllerGetCapabilities(context.Context, *ControllerGetCapabilitiesRequest) (*ControllerGetCapabilitiesResponse, error)
-*/
+	if len(req.GetSourceVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	} else if len(req.GetName()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Name must be provided")
+	}
+
+	// Check if the snapshot with this name already exists
+	v, err := util.VolumeFromName(o.driver, req.GetName())
+	if err == nil {
+		// Verify the parent is the same
+		if req.GetSourceVolumeId() != v.GetSource().GetParent() {
+			return nil, status.Error(codes.AlreadyExists, "Requested snapshot already exists for another source volume id")
+		}
+
+		// Return current snapshot info
+		createdAt, err := ptypes.Timestamp(v.GetCtime())
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Failed to get time snapshot was created: %v", err)
+		}
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				Id:             v.GetId(),
+				SourceVolumeId: v.GetSource().GetParent(),
+				CreatedAt:      createdAt.Unix(),
+				Status: &csi.SnapshotStatus{
+					// This means that we are not uploading our snapshot
+					// We may add support flow cloud snaps in future patches
+					Type: csi.SnapshotStatus_READY,
+				},
+			},
+		}, nil
+	}
+
+	// Get any labels passed in by the CO
+	_, locator, _, err := o.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Unable to get parameters: %v", err)
+	}
+
+	// Create snapshot
+	readonly := true
+	snapshotID, err := o.driver.Snapshot(req.GetSourceVolumeId(), readonly, &api.VolumeLocator{
+		Name:         req.GetName(),
+		VolumeLabels: locator.GetVolumeLabels(),
+	}, false)
```

Comment:
We don't want the backend implementation to retry. 
https://github.com/libopenstorage/stork/blob/master/drivers/volume/portworx/portworx.go#L611 is how stork uses it.

The PX implementation will retry and eventually create a snapshot with same name. When the caller reinvokes the request with same name, PX will fail this with already exists error.

cc: @disrani-px

Reply:
It retries only if noRetry is set to false.

Since we are allowing existing snapshots, I think we are fine here.